updated for csproj/net 2.0

mairaw opened the pull request:
I've noticed this topic had still references to project.json and didn't contain the latest TFMs.

Related to #2091 

[Internal Preview URL](https://review.docs.microsoft.com/en-us/dotnet/core/porting/third-party-deps?branch=pr-en-us-2860)

Fixes #459 

edit by @billwagner Added "fixes" comment based on @mairaw comment on that issue.

mairaw committed:
updated for csproj/net 2.0

cartermp reviewed (approved):
Looks good, though I definitely wouldn't merge this without a :shipit: from @terrajobst

mairaw commented:
Ok, I'll wait for @terrajobst's review! Thanks @cartermp!

terrajobst commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,35 @@
 ## Prerequisites
 
-This article will assume you are using Windows and Visual Studio, and that you have code which runs on the .NET Framework today.
+This article assumes you're using Windows and Visual Studio, and that you have code that currently runs on the .NET Framework.
```
It seems a bit weird to mention Windows here because it makes it sound we're only talking about how you can port to .NET Core running on Windows. We really mean that we assume the customer is *starting* on Windows with .NET Framework. The end result might run on Linux.

mairaw replied:
Makes sense. I'll change it to make it more clear.

terrajobst commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,35 @@
+## Analyzing NuGet packages
 
-Analyzing NuGet packages for portability is very easy.  Because a NuGet package is itself a set of folders which contain platform-specific assemblies, all you have to do is check to see if there is a folder which contains a .NET Core assembly.
+Analyzing NuGet packages for portability is easy. A NuGet package is itself a set of folders that contain platform-specific assemblies. So, what you have to do is check to see if there's a folder that contains a compatible assembly inside the package.
```
While true, we need to call out the .NET Framework compatibility mode we have now. See my upcoming blog post on .NET Standard 2.0. It has a detailed walkthrough on how this works. In essence, it allows people to consume .NET Framework libraries when targeting .NET Standard 2.0 (or any .NET implementation of it). They get a warning telling them that library might not work, but they are no longer blocked as they are today. After testing, they can suppress the warning.

mairaw replied:
I saw that. Will add.

terrajobst reviewed (changes requested):
We need to call out the new .NET Framework compatibility mode that comes with .NET Standard 2.0.

mairaw committed:
feedback

mairaw commented:
Sorry for the super mega delay on addressing the feedback here @terrajobst. Please take a look.

I'm also wondering if we wanna keep the portable TFMs and the pre-release TFMs in this topic. Thoughts?

svick commented on docs/core/porting/third-party-deps.md:
```diff
@@ -42,14 +35,19 @@ netstandard1.3
-These are the Target Framework Monikers (TFM) which map to versions of the [.NET Standard](../../standard/net-standard.md) and traditional Portable Class Library (PCL) profiles which are compatible with .NET Core.  Note that `netcoreapp1.0`, while compatible, is for applications and not libraries.  Although there's nothing wrong with using a library which is `netcoreapp1.0`-based, that library ma ...
+These values are the [Target Framework Monikers (TFM)](../../standard/frameworks.md) that map to versions of the [.NET Standard](../../standard/net-standard.md), .NET Core, and traditional Portable Class Library (PCL) profiles that are compatible with .NET Core. Note that `netcoreapp`, while compatible, is for .NET Core projects only. Although there's nothing wrong with using a .NET Core library  ...
+
+Alternatively, you can see the TFMs that each package supports on [nuget.org](https://www.nuget.org/) under the **Dependencies** section of the package page.
```
> The easiest way to inspect NuGet Package folders is to use the NuGet Package Explorer tool. After downloading it, use the following steps to see the folder names:
> …
> Alternatively, you can see the TFMs that each package supports on nuget.org under the **Dependencies** section of the package page.

How is installing a Windows-only tool and following a 5-step process easier than checking a website? Shouldn't it be the other way around, i.e. talk about nuget.org first and mention NPE as the alternative?

mairaw replied:
I think it's a good idea. I think the issue is that NuGet doesn't always have that info. So instead of easiest, we could say that the tool is the more reliable way to check or something. See the package we use here as an example: https://www.nuget.org/packages/Huitian.PowerCollections/ and you'll see the dependencies are not listed

svick commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,31 @@
+Analyzing NuGet packages for portability is easy. A NuGet package is itself a set of folders that contain platform-specific assemblies. So, what you have to do is check to see if there's a folder that contains a compatible assembly inside the package.
 
-This article will assume you are using Windows and Visual Studio, and that you have code which runs on the .NET Framework today.
+The easiest way to inspect NuGet Package folders is to use the [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer) tool. After downloading it, use the following steps to see the folder names:
```
Nit: the primary way of distributing NPE is the Windows Store, which means you don't really download it, you install it. So I would say:

> After installing it, use the following steps to see the folder names:

svick commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,31 +61,53 @@ dotnet5.4
 
 > [!NOTE]
-> To use a package targeting a traditional PCL or pre-release .NET Core target, you must use the `imports` directive in your `project.json` file.
+> To use a package targeting a traditional PCL or pre-release .NET Core target, you must use the `PackageTargetFallback` MSBuild element in your project file.
```
How do I do that? Would a link to some documentation about this (possibly https://docs.microsoft.com/nuget/schema/msbuild-targets#packagetargetfallback) be useful here?

mairaw replied:
That's a good suggestion @svick, while we still don't have our own MSBuild reference topic for .NET Core.

svick commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,31 +61,53 @@ dotnet5.4
+
+`NU1701: Package ‘Huitian.PowerCollections 1.0.0’ was restored using ‘.NETFramework,Version=v4.6.1’ instead of the project target framework ‘.NETStandard,Version=v2.0’. This package may not be fully compatible with your project.`
+
+That warning is displayed when you add the package and every time you build to make sure you test that package with your project. If your project is working as expected, you can suppress that warning by editing the package properties on Visual Studio or by manually editing the project file on your favorite code editor.
```
I think "on Visual Studio" and "on your favorite code editor" should use "in" instead of "on".

mairaw replied:
I think you're right. Prepositions are my Achilles heel. @rpetrusha to confirm. 😄

svick commented on docs/core/porting/third-party-deps.md:
````diff
@@ -63,31 +61,53 @@ dotnet5.4
+</ItemGroup>
+```
+
+For more information on how to suppress compiler warnings in Visual Studio, see [How to: Suppress Compiler Warnings](/visualstudio/ide/how-to-suppress-compiler-warnings).
````
Should this link directly to the section about NuGet packages (https://docs.microsoft.com/visualstudio/ide/how-to-suppress-compiler-warnings#suppressing-warnings-for-nuget-packages)?

mairaw replied:
Yes! That section didn't exist when I added this link. 😄

mairaw committed:
feedback and another review pass

mairaw commented:
@rpetrusha @BillWagner or @JRAlexander can one of you also do an UE review? Thanks! 

I think this one should be ready to go now (finally).

mairaw committed:
remove dupe info

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,40 @@
+There are a few ways to accomplish that:
 
-This article will assume you are using Windows and Visual Studio, and that you have code which runs on the .NET Framework today.
+* Using the NuGet Package Explorer app (most reliable method).
```
These could be links to the relevant sections. I think that that would enhance the readability of the topic.
also, most --> the most

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,40 @@
 
-## Prerequisites
+If you're referencing NuGet packages in your project, you need to verify if they're compatible with .NET Core.
+There are a few ways to accomplish that:
```
a few ways --> two ways

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,40 @@
 
-This article will assume you are using Windows and Visual Studio, and that you have code which runs on the .NET Framework today.
+* Using the NuGet Package Explorer app (most reliable method).
+* Using [nuget.org](https://www.nuget.org/) site.
```
Using --> Using the

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,40 @@
+* Using [nuget.org](https://www.nuget.org/) site.
 
-## Analyzing NuGet Packages
+After analyzing the packages, if they're not compatible with .NET Core and only target .NET Framework, you can check if the [.NET Framework compatibility mode](#net-framework-compatibility-mode) can help on your porting process.
```
on --> with

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -1,38 +1,40 @@
+### Analyze NuGet packages using NuGet Package Explorer
 
-Inspecting NuGet Package folders is easiest with the [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer) tool.  Here's how to do it.
+A NuGet package is itself a set of folders that contain platform-specific assemblies. So, you need to check if there's a folder that contains a compatible assembly inside the package.
```
no comma after So

rpetrusha commented on docs/core/porting/third-party-deps.md:
````diff
@@ -42,14 +44,17 @@ netstandard1.3
 ```
 
-These are the Target Framework Monikers (TFM) which map to versions of the [.NET Standard](../../standard/net-standard.md) and traditional Portable Class Library (PCL) profiles which are compatible with .NET Core.  Note that `netcoreapp1.0`, while compatible, is for applications and not libraries.  Although there's nothing wrong with using a library which is `netcoreapp1.0`-based, that library ma ...
+These values are the [Target Framework Monikers (TFM)](../../standard/frameworks.md) that map to versions of the [.NET Standard](../../standard/net-standard.md), .NET Core, and traditional Portable Class Library (PCL) profiles that are compatible with .NET Core. Note that `netcoreapp`, while compatible, is for .NET Core projects only and not for .NET Standard. Although there's nothing wrong with  ...
````
(TFM) --> (TFMs)

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,30 +68,59 @@ dotnet5.4
 > [!NOTE]
-> To use a package targeting a traditional PCL or pre-release .NET Core target, you must use the `imports` directive in your `project.json` file.
+> To use a package targeting a traditional PCL or pre-release .NET Core target, you must use the `PackageTargetFallback` MSBuild element in your project file.
+> For more information about this MSBuild element, see [`PackageTargetFallback`](../tools/csproj.md#packagetargetfallback).
```
no paragraph break.

mairaw replied:
There's no paragraph break here (I'd need an extra line in Markdown to create that, so keeping as-is)

![image](https://user-images.githubusercontent.com/12971179/35716205-1f36b98c-078c-11e8-9ef4-415e1c1f5934.png)

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,30 +68,59 @@ dotnet5.4
+
+Starting with .NET Standard 2.0, the .NET Framework compatibility mode was introduced. This compatibility mode allows .NET Standard and .NET Core projects to reference .NET Framework libraries. Referencing .NET Framework libraries doesn't work for all projects, such as if the library uses Windows Presentation Foundation (WPF) APIs, but it does unblock many porting scenarios.
+
+When you reference NuGet packages that target the .NET Framework on your project, such as [Huitian.PowerCollections](https://www.nuget.org/packages/Huitian.PowerCollections), you get a package fallback warning ([NU1701](/nuget/reference/errors-and-warnings#nu1701)) similar to the following example:
```
on --> in

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,30 +68,59 @@ dotnet5.4
+3. You can search for another package that runs on .NET Core that accomplishes the same task as the package you were using.
 4. You can attempt to write the code the package was doing yourself.
-5. You could eliminate the dependency on the package by changing the functionality of your app, at least until a compatible version of the package becomes available.
+5. You could eliminate the dependency on the package by changing the functionality of your app. At least, until a compatible version of the package becomes available.
```
"At least..." shouldn't be a separate sentence.

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,30 +68,59 @@ dotnet5.4
+If you can't resolve your issue with any of the above, you may have to port to .NET Core at a later date.
 
-The .NET Team would like to know which libraries are the most important to support next with .NET Core. You can also send us mail at dotnet@microsoft.com about the libraries you'd like to use.
+The .NET Team would like to know which libraries are the most important to support next with .NET Core. You can send an email to dotnet@microsoft.com about the libraries you'd like to use.
```
"next" is best eliminated.

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -63,30 +68,59 @@ dotnet5.4
+## Analyze dependencies that aren't NuGet packages
 
-You may have a dependency that isn't a NuGet package, such as a DLL in the filesystem.  The only way to determine the portability of that dependency is to run the [ApiPort tool](https://github.com/Microsoft/dotnet-apiport/blob/master/docs/HowTo/).
+You may have a dependency that isn't a NuGet package, such as a DLL in the filesystem. The only way to determine the portability of that dependency is to run the [ApiPort tool](https://github.com/Microsoft/dotnet-apiport/blob/master/docs/HowTo/).
```
filesystem --> filesystem
Add a parenthetical description of what the APIPort tool does.

mairaw replied:
I think you mean to change it to file system?

rpetrusha reviewed (changes requested):
You've done a good job with the revisions, @mairaw. These are mostly nits, with a few other comments.

mairaw committed:
feedback

mairaw reviewed (commented):
Thanks for your thorough review @rpetrusha. I appreciate it!

mairaw committed:
Merge branch 'master' into port3deps

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -65,30 +70,59 @@ dotnet5.4
+
+Alternatively, you can see the TFMs that each package supports on [nuget.org](https://www.nuget.org/) under the **Dependencies** section of the package page.
+
+Although this is an easier method to verify the compatibility, the **Dependencies** information is not available on the site for all packages.
```
nit: the **Dependencies** --> **Dependencies**

rpetrusha commented on docs/core/porting/third-party-deps.md:
```diff
@@ -65,30 +70,59 @@ dotnet5.4
 5. You could eliminate the dependency on the package by changing the functionality of your app, at least until a compatible version of the package becomes available.
 
-Please remember that open source project maintainers and NuGet package publishers are often volunteers who contribute because they care about a given domain, do it for free, and often have a different daytime job. If you do reach out, you might start with a positive statement about the library before asking about .NET Core support.
+Remember that open-source project maintainers and NuGet package publishers are often volunteers. They contribute because they care about a given domain, do it for free, and often have a different daytime job. So, be mindful of that when contacting them to ask for .NET Core support.
```
nit: no comma after So

rpetrusha reviewed (approved):
Three nits, @mairaw, You can decide whether you want to address them before you merge.

BillWagner commented on docs/core/porting/third-party-deps.md:
````diff
@@ -44,14 +46,17 @@ netstandard1.3
 ```
 
-These are the Target Framework Monikers (TFM) which map to versions of the [.NET Standard](../../standard/net-standard.md) and traditional Portable Class Library (PCL) profiles which are compatible with .NET Core.  Note that `netcoreapp1.0`, while compatible, is for applications and not libraries.  Although there's nothing wrong with using a library which is `netcoreapp1.0`-based, that library ma ...
+These values are the [Target Framework Monikers (TFMs)](../../standard/frameworks.md) that map to versions of the [.NET Standard](../../standard/net-standard.md), .NET Core, and traditional Portable Class Library (PCL) profiles that are compatible with .NET Core. Note that `netcoreapp`, while compatible, is for .NET Core projects only and not for .NET Standard. Although there's nothing wrong with ...
````
The last two sentences in this paragraph are really important for customers to understand. Is there a way to make them stand out more? Make them their own paragraph, or highlight them in some other way?

mairaw replied:
Let me know what you think of the changes @BillWagner

BillWagner reviewed (approved):
This looks great @mairaw   

I only had one question, and I don't have a great way to solve it, so :shipit: when ready.

mairaw committed:
feedback

mairaw commented:
I've addressed all the remaining feedback. @terrajobst please take a look to see if now you can approve the PR. 

My final question is: do we still wanna list PCL and pre-release TFMs in this topic? 

Thanks!

terrajobst commented:
> My final question is: do we still wanna list PCL and pre-release TFMs in this topic?

I think it's topical here as it's all about compat with the past when migrating.

mairaw commented:
Cool. Thanks @terrajobst. 🎆
